src/interactive-auth.js

```diff
@@ -88,21 +89,30 @@ const MSISDN_STAGE_TYPE = "m.login.msisdn";
  * @param {string?} opts.emailSid If returning from having completed m.login.email.identity
  *     auth, the sid for the email verification session.
  *
+ * @param {function?} opts.requestEmailToken A function that takes the email address (string),
+ *     clientSecret (string), attempt number (int) and sessionId (string) and calls the
+ *     relevant requestToken function and returns the promise returned by that function.
+ *     If the resulting promise rejects, the rejection will propagate through to the
+ *     attemptAuth promise.
+ *
  */
 function InteractiveAuth(opts) {
     this._matrixClient = opts.matrixClient;
     this._data = opts.authData || {};
     this._requestCallback = opts.doRequest;
     // startAuthStage included for backwards compat
     this._stateUpdatedCallback = opts.stateUpdated || opts.startAuthStage;
-    this._completionDeferred = null;
+    this._resolveFunc = null;
+    this._rejectFunc = null;
     this._inputs = opts.inputs || {};
+    this._requestEmailTokenCallback = opts.requestEmailToken;
```

Comment:
The callback provided by the react-sdk PR calls straight back into the JS-SDK. Curious why you opted for not doing that straight in here?

Reply:
It's because there is a requestToken function for each endpoint which all have slightly different semantics (ie. the one for registration errors if the email address is already registered, the one for password reset errors if the email is *not* registered) and interactive-auth doesn't know what endpoint you're calling, so it doesn't know what requestToken method to use.